File: server/graph/model/user_request.go
package model

import "time"

type UserRequest struct {
	ID               string            `json:"id" gorm:"type:uuid;primarykey;default:gen_random_uuid()"`
	AnimalID         string            `json:"-"`
	Animal           ShelterAnimal     `json:"animal" validate:"-" gorm:"foreignKey:AnimalID;references:ID;OnDelete:SET NULL"`
	UserID           string            `json:"-"`
	User             User              `json:"user" validate:"-" gorm:"foreignKey:UserID;references:ID;OnDelete:SET NULL"`
	RequestType      UserRequestType   `json:"requestType"`
	RequestStatus    UserRequestStatus `json:"requestStatus" gorm:"default:Pending"`
	FromDate         *time.Time        `json:"fromDate"`
	ToDate           *time.Time        `json:"toDate"`
	ApprovedByUserID *string           `json:"-"`
	ApprovedBy       *User             `json:"approvedByUser" validate:"-" gorm:"foreignKey:ApprovedByUserID;references:ID;OnDelete:SET NULL;default:null"`
	AddDatetime      time.Time         `json:"addDatetime" gorm:"not null;default:current_timestamp"`
}

File: app/lib/models/shelter/user_request.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:furry_nebula/graphql/fragments/__generated__/user_request_fragment.data.gql.dart';
import 'package:furry_nebula/models/shelter/shelter_animal.dart';
import 'package:furry_nebula/models/shelter/user_request_status.dart';
import 'package:furry_nebula/models/shelter/user_request_type.dart';
import 'package:furry_nebula/models/user/user.dart';

part 'user_request.freezed.dart';

@freezed
class UserRequest with _$UserRequest {
  const factory UserRequest({
    required String id,
    required UserRequestType requestType,
    required UserRequestStatus requestStatus,
    required User user,
    required ShelterAnimal animal,
    User? approvedBy,
    DateTime? fromDate,
    DateTime? toDate,
  }) = _UserRequest;

  factory UserRequest.fromFragment(GUserRequestFragment fragment) =>
      UserRequest(
        id: fragment.id,
        requestType: UserRequestType.fromGUserRequestType(fragment.requestType)
            ?? UserRequestType.accommodation,
        requestStatus: UserRequestStatus.fromGUserRequestStatus(fragment.requestStatus)
            ?? UserRequestStatus.pending,
        user: User.fromFragment(fragment.user),
        animal: ShelterAnimal.fromFragment(fragment.animal),
        approvedBy: fragment.approvedBy != null
            ? User.fromFragment(fragment.approvedBy!)
            : null,
        fromDate: fragment.fromDate,
        toDate: fragment.toDate,
      );

  const UserRequest._();
}
